firmware/library/L1_Drivers/pwm.hpp

```diff
@@ -1,13 +1,176 @@
+        if (pwm_pin_.GetPort() == 2)
+        {
+            port_ = 1;
+            channel_ = static_cast<uint8_t>(pwm_pin_.GetPin() + 1);
+        }
+    }
+
+    Pwm(const PinConfigureInterface & gpio_):
+        pwm_(gpio_),
+        pwm_pin_(PinConfigure::CreateInactivePin()),
+        port_(pwm_pin_.GetPort()),
+        channel_(pwm_pin_.GetPin()),
+        duty_cycle(0)
+    {
+    }
+
+
+    ~Pwm()
+    {
+        pwm_pin_.SetPinFunction(kGpioFunction);
+    }
+
+    void Initialize(uint32_t frequency_hz = 1'000) override
+    {
+        // TODO: Remove when sysclock complete.
+        // DIRECT MANIUPULATION OF SYSTEM CLOCK REGISTERS
+        // WILL REPLACE WITH FUNCTIONS LATER
+        // Enables PWM1 power/clock control bit
+        LPC_SC->PCONP |= SysClk::kActivatePwm1;
+        // END OF DIRECT MANIPULATION
+
+        match_value = config::kSystemClockRate / frequency_hz;
+
+        // Resets PWMTC on Match with MR0
+        LPC_PWM1->MCR |= PwmConfigure::kResetMr0;
+        LPC_PWM1->MR0 = match_value;
+
+        // Enables PWM TC and PC for counting and enables PWM mode
+        LPC_PWM1->TCR = PwmConfigure::kCounterEnable | PwmConfigure::kPwmEnable;
+        LPC_PWM1->CTCR &= PwmConfigure::kTimerMode | PwmConfigure::kUsePwmCap0;
+
+
+        // Enables PWM[channel] output
+        LPC_PWM1->PCR |= PwmOutputEnable(channel_);
+
+        pwm_pin_.SetPinFunction(kPwmFunction);
+    }
+
+    void SetDutyCycle(float percentage) override
+    {
+
+        if (percentage < 0 || percentage > 100)
+        {
+            // TODO: Replace with an assert 
+            return;
+        }
+        duty_cycle = static_cast<uint32_t>((percentage * static_cast<float>(match_value)) / 100);
+
+        // Pointer to point to address of Match Register
+        volatile uint32_t * match_register_selector;
+
+        // Have the pointer point to address of MR
+        match_register_selector = match_register_table[channel_];
+        // Set the value of the MR
+        *match_register_selector = duty_cycle;
+        LPC_PWM1->LER |= (1 << channel_);
+    }
+
+    float GetDutyCycle() override
+    {
+        return static_cast<float>(duty_cycle / match_value);
+    }
+
+    void SetFrequency(uint32_t frequency_hz) override
+    {
+        // Disables PWM mode; this will reset all counters to 0
+        // And allow us to update MR0
+        LPC_PWM1->TCR &= ~PwmConfigure::kPwmEnable;
+
+        match_value = config::kSystemClockRate / frequency_hz;
+        LPC_PWM1->MR0 = match_value;
+
+        // Re-enable PWM mode
+        LPC_PWM1->TCR |= PwmConfigure::kPwmEnable;
+    }
+
+    uint32_t GetFrequency() override
+    {
+        return config::kSystemClockRate / match_value;
+    }
+
+    inline uint8_t PwmOutputEnable(uint8_t channel)
+    {
+        return static_cast<uint8_t>(1 <<  (channel + 8));
+    }
+
+ private:
+    const PinConfigureInterface & pwm_;
+
+    PinConfigure pwm_pin_;
+
+    static uint32_t match_value;
```

Comment:
Rather than keeping this local to the class, you could use a constexpr * match_value. This is not thread safe. The source of truth about the match value, should be the register not a variable.